ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs

```diff
@@ -909,21 +936,33 @@ jobDemoteEstablishedPeer PeerSelectionActions {
                                  (Set.size $ EstablishedPeers.toSet establishedPeers
                                              `Set.intersection`
                                              bigLedgerPeersSet)
-                                 peeraddr e]
+                                 peeraddr e']
                           else [TraceDemoteWarmFailed
                                  targetNumberOfEstablishedPeers
                                  (Set.size $ EstablishedPeers.toSet establishedPeers
                                       Set.\\ bigLedgerPeersSet)
-                                 peeraddr e],
+                                 peeraddr e'],
           decisionState = st {
-                            inProgressDemoteWarm = inProgressDemoteWarm'
+                            inProgressDemoteWarm = inProgressDemoteWarm',
+                            establishedPeers     = establishedPeers'
                           },
           decisionJobs  = []
       }
 
     job :: m (Completion m extraState extraDebugState extraFlags extraPeers peeraddr peerconn)
     job = do
-      closePeerConnection peerconn
+      -- cf. issue #5111 - we wait for the connection to be forgotten by the CM
+      -- before we return. Otherwise, if the same peer is promoted too quickly by
+      -- eg. churn, the connection will fail.
+      res <-
+        timeout c_PEER_DEMOTION_TIMEOUT do
+          closePeerConnection peerconn
```

Comment:
`closePeerConnection` has it's own timeout (set to `120s`); it will be cearer if we just wrap the atomic action below in its own timeout.